### PR TITLE
Initial implementation for RDF-based tests QueryEvaluationTest

### PR DESCRIPTION
**Pull request by Mec-iS**
From #248

This is the first draft for implementing automated RDF tests.
The first batch of test from `RDF-tests` (basic) can be run with `pytest tests/rdf_tests/test_rdf_basic.py -k test_rdf_runner -s` from the project directory; tests are in `tests/rdf_tests/dat`. For the moment the only assertion implemented is a check on the length of returned results. This is already interesting as some kg query outputs return more or less bindings compared to the expected ones.

The script should work also on `oxigraph-tests` but it doesn't, it seems the `read_manifest` file provided with `rdflib` cannot parse the oxigraph manifest, maybe a discrepancy in the XML structure? @Tpt please provide some feedback.

Currently the tests are cut-pasted.

Tests with anomalies:
* Running basic/Non-matching triple pattern ERROR: Non-matching triple pattern resulted in object of type 'NoneType' has no len()
* Running basic/Basic - Prefix/Base 1 Basic - Prefix/Base 1 resulted in False
* Running basic/Basic - Var 1 Basic - Var 1 resulted in False
* Running basic/Basic - Var 2 Basic - Var 2 resulted in False
* Running basic/Basic - Term 6 Basic - Term 6 resulted in False
* Running basic/Basic - Term 7 ERROR: Basic - Term 7 resulted in Expected {SelectQuery | ConstructQuery | DescribeQuery | AskQuery}, found '.'  (at char 140), (line:5, col:23)

The ones resulting in "False" mean length discrepancies between expected and actual output, the one with errors are exceptions raised.

### CODE REVIEW

File: tests/rdf_tests/rdflib_tools.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+        for col in g.objects(m, MF.entries):
+            e: URIRef
+            for e in g.items(col):
+
+                approved = (
+                    (e, DAWG.approval, DAWG.Approved) in g
+                    or (e, DAWG.approval, DAWG.NotClassified) in g
+                    or (e, RDFT.approval, RDFT.Approved) in g
+                )
```
Comment (Tpt) [inline tests/rdf_tests/rdflib_tools.py:90]: Oxigraph tests are not from the W3C working group so the DAWG.approval property is not set.

File: tests/rdf_tests/rdflib_tools.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+                # run legacy tests with no approval set
+                if legacy:
+                    approved |= (e, DAWG.approval, None) not in g and (
+                        e,
+                        RDFT.approval,
+                        None,
+                    ) not in g
+
+                if not approved:
```
Comment (Tpt) [inline tests/rdf_tests/rdflib_tools.py:103]: Oxigraph tests are not from the W3C working group so the DAWG.approval property is not set, leading to `approved = False`. This explains why Oxigraph tests are ignored.

File: tests/rdf_tests/dat/oxigraph-tests/sparql/order_terms.rq [hunk 3]
```diff
@@ -0,0 +1,3 @@
+SELECT ?o WHERE {
+    VALUES ?o { 4.0 "v" "u"@en "a" "d"@de "c"@en "b" "foo"^^<http://example.com/h> 1 "2021-30-30"^^<http://www.w3.org/2001/XMLSchema#date> "2020-30-29T00:00:00"^^<http://www.w3.org/2001/XMLSchema#dateTime> -1e3 }
+} ORDER BY ?o
```
Comment (Tpt) [inline tests/rdf_tests/dat/oxigraph-tests/sparql/order_terms.rq:3]: This ordering test is testing something implementation defined. I am not sure it makes sense to keep it here.

File: tests/rdf_tests/test_rdf_basic.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    kg = kglab.KnowledgeGraph(
+        name = "RDF Tests Graph",
+        base_uri = "http://example/",
+        namespaces = namespaces,
+    )
+
+    #
+    # load test data in the graph
+    #
+    kg.load_rdf(t.data)
```
Comment (Tpt) [inline tests/rdf_tests/test_rdf_basic.py:45]: `t.data` might be `None` if no input file is used.

File: tests/rdf_tests/test_rdf_basic.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    ic(manifest_file)
+
+    yield read_manifest(manifest_file)
+
+def clean_filepath(filepath):
+    """ Clean paths loaded from RDF manifest """
+    return str(Path(str(filepath))).replace("file:", "")
+    
+
+def run_test(t):
```
Comment (Tpt) [inline tests/rdf_tests/test_rdf_basic.py:29]: Beware, you support here only `mf:QueryEvaluationTest` and not the other kind of tests (updates, syntax...).

**Review by Tpt — COMMENTED** [review]
Thank you so much for moving this forward!

I have found why the Oxigraph tests are ignored (see my inline comment in`rdflib_tools.py`). I also allowed myself to comment about some problems I found while quickly trying the code.

**Comment by Tpt** [comment]
By the way, Oxigraph does not pass all official SPARQL tests. This is mostly due to the fact that Oxigraph storage normalizes literals like numbers. For example `"01"^^xsd:integer` and `"1"^^xsd:integer` are considered to be the same. This leads to the failure of some tests taking care of duplicates. The list of failing test [is here](https://github.com/oxigraph/oxigraph/blob/cbccdfba867204ce4b20b6fc16e37a30719f90eb/testsuite/tests/sparql.rs#L36).

**Comment by Mec-iS** [comment]
Is there any straightforward way of making `rdflib` SPARQL query to return serialised data (ttl or xml) instead of the row iterator?

**Comment by Tpt** [comment]
> Is there any straightforward way of making `rdflib` SPARQL query to return serialised data (ttl or xml) instead of the row iterator?
> 
> or what is the given way of testing the result of an rdflib query to a given ttl file? @ceteri

RDFlib provides [a parser for SPARQL results encoded in RDF](https://github.com/RDFLib/rdflib/blob/03c5e79298b35f29087ff82002becaaba0e3a656/rdflib/plugins/sparql/results/rdfresults.py#L8).
Then it uses [this function](https://github.com/RDFLib/rdflib/blob/03c5e79298b35f29087ff82002becaaba0e3a656/test/test_dawg.py#L126) to check if results sets are compatible.

I tried in Oxigraph to compare results set encoded in RDF using the graph isomorphism algorithm. It was very slow because results sets encoded in RDF contains a lot of blank nodes only connected to other blank nodes, making the hash based alogirthms not very efficient...

**Comment by Mec-iS** [comment]
In some tests `kglab` fails, most of them are "Python recursion limit exceeded".

**Comment by ceteri** [comment]
@Mec-iS this is excellent, and @Tpt thank you kindly!

**Comment by ceteri** [comment]
In terms of _graph isomorphism algorithms_, I wish there was more open source available for graph sketch algorithms. That might help with the costs. I've only found one <https://github.com/kenkoooo/graph-sketch-fractality> although it's based on a CLI and not quite the similarity measures that we'd need.

**Pull request merged**